stdlib/public/runtime/StringComparison.cpp

```diff
@@ -0,0 +1,161 @@
+} // end anonymous namespace
+
+// *TODO* use a template to make this autoupgrade to use AVX when -mavx is
+// *passed in.
+static int swift_stdlib_findDiffIdx_UInt8UInt16_vector(uint8_t *left,
+                                                       uint16_t *right,
+                                                       int initialCount) {
+  int count = initialCount;
+  auto *leftVPtr = reinterpret_cast<simd_packed_uchar16 *>(left);
+  auto *rightVPtr = reinterpret_cast<simd_packed_ushort8 *>(right);
+
+  // Our main loop is going to be processing 64 elements at a time. This can be
+  // seen by us processing 4 vectors of 16 elements and 8 vector of 8 elements.
+  //
+  // TODO: Do page size blocking.
+  while (count >= ELTS_PER_ROUND) {
+    auto leftV0 = leftVPtr[0];
+    auto leftV1 = leftVPtr[1];
+    auto leftV2 = leftVPtr[2];
+    auto leftV3 = leftVPtr[3];
+
+    auto rightV0 = rightVPtr[0];
+    auto rightV1 = rightVPtr[1];
+    auto rightV2 = rightVPtr[2];
+    auto rightV3 = rightVPtr[3];
+    auto rightV4 = rightVPtr[4];
+    auto rightV5 = rightVPtr[5];
+    auto rightV6 = rightVPtr[6];
+    auto rightV7 = rightVPtr[7];
+
+    // Then compare the upper/lower halves of our v16u8 vectors with v8u16
+    // vectors.
+    auto V0 = performComparisonLow(leftV0, rightV0);
+    auto V1 = performComparisonHigh(leftV0, rightV1);
+    auto V2 = performComparisonLow(leftV1, rightV2);
+    auto V3 = performComparisonHigh(leftV1, rightV3);
+    auto V4 = performComparisonLow(leftV2, rightV4);
+    auto V5 = performComparisonHigh(leftV2, rightV5);
+    auto V6 = performComparisonLow(leftV3, rightV6);
+    auto V7 = performComparisonHigh(leftV3, rightV7);
+
+    // Then reduce over vectors to see if any of our vectors were found to
+    // compare not equal.
+    V0 |= V1;
+    V2 |= V3;
+    V4 |= V5;
+    V6 |= V7;
+
+    V0 |= V2;
+    V4 |= V6;
+
+    V0 |= V4;
+
+    // Finally if any of our vectors were found to compare not equal, then back
+    // track and find the actual index. This should almost always predict false,
+    // so we should only mispredict on the last iteration.
+    if (simd_any(V0)) {
+      // We are assumed to return the offset from the beginning of the array
+      // where the error happened. We have already processed (initialCount -
+      // count) at this point. So we just need to add in the index found by
+      // findDiffIdxSmall.
+      unsigned offset = initialCount - count;
+      offset += findDiffIdxSmall(reinterpret_cast<uint8_t *>(leftVPtr),
+                                 reinterpret_cast<uint16_t *>(rightVPtr),
+                                 ELTS_PER_ROUND);
+      return offset;
+    }
+
+    count -= 64;
+    leftVPtr += 4;
+    rightVPtr += 8;
+  }
+
+  return initialCount;
+}
+
+#endif
+
+static int findDiffIdxSmall(uint8_t *left, uint16_t *right, int count) {
+  for (int i = 0; i != count; ++i) {
+    if (uint16_t(left[i]) == right[i]) {
```

Comment:
This ought to be !=, right?

Reply:
Yes. I fixed that locally.

Reply:
What happened here is I copied the swift code but it was using a guard and I didn't invert.